plugins/flashrom-i2c/fu-plugin-flashrom-i2c.c

```diff
@@ -0,0 +1,198 @@
+	quirk_programmer_name = fu_plugin_lookup_quirk_by_id (
+		plugin, quirk_key, PROGRAMMER_NAME);
+	quirk_device_name = fu_plugin_lookup_quirk_by_id (
+		plugin, quirk_key, DEVICE_NAME);
+	quirk_device_protocol = fu_plugin_lookup_quirk_by_id (
+		plugin, quirk_key, DEVICE_PROTOCOL);
+	/* Add devices with quirk configuration only. */
+	if (quirk_programmer_name == NULL)
+		return TRUE;
+
+	device_path = g_build_filename (
+		i2c_device_dir, i2c_name, NULL);
+	bus_no = get_i2c_bus_number_from_path (device_path, error);
+	if (bus_no == -1) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "%s failed to get bus number for " \
+			     "device under %s.",
+			     __func__,
+			     device_path);
+		return FALSE;
+	}
+
+	device_vid = g_strndup (hw_id, HID_LENGTH / 2);
+	device_pid = g_strndup (&hw_id[HID_LENGTH / 2], HID_LENGTH / 2);
+	device_guid = get_i2c_device_guid (device_vid, device_pid);
+	if (device_guid == NULL) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "%s failed to generate guid.",
+			     __func__);
+		return FALSE;
+	}
+
+	vendor_id = g_strdup_printf ("I2C:%s", device_vid);
+	physical_id = g_strdup_printf ("DEVNAME=%s", device_path);
+	dev = g_object_new (FU_TYPE_I2C_DEVICE, NULL);
+
+	fu_device_add_guid (FU_DEVICE (dev), device_guid);
+	fu_device_set_vendor (FU_DEVICE (dev), fu_plugin_get_dmi_value (
+		plugin, FU_HWIDS_KEY_MANUFACTURER));
+	fu_device_set_vendor_id (FU_DEVICE (dev), vendor_id);
+	fu_device_set_version_format (FU_DEVICE (dev),
+				      FWUPD_VERSION_FORMAT_PAIR);
+	/* TODO(b/154178623): Get the real version number using flashrom. */
+	fu_device_set_version (FU_DEVICE (dev), "0.0");
+	fu_device_set_name (FU_DEVICE (dev), quirk_device_name);
+	fu_device_set_protocol (FU_DEVICE (dev), quirk_device_protocol);
+	fu_device_set_physical_id (FU_DEVICE (dev), physical_id);
+	fu_device_set_metadata_integer (FU_DEVICE (dev), PORT_NAME, bus_no);
+	fu_device_set_metadata (FU_DEVICE (dev), PROGRAMMER_NAME,
+		quirk_programmer_name);
+	fu_plugin_device_add (plugin, FU_DEVICE (dev));
+	return TRUE;
+}
+
+void fu_plugin_init (FuPlugin *plugin)
+{
+	fu_plugin_set_build_hash (plugin, FU_BUILD_HASH);
+	fu_plugin_set_device_gtype (plugin, FU_TYPE_I2C_DEVICE);
+}
+
+gboolean fu_plugin_coldplug (FuPlugin *plugin, GError **error)
+{
+	g_autofree gchar *i2c_device_dir =
+		fu_common_get_path (FU_PATH_KIND_I2C_DEVICES);
+	GDir *dir;
+	const gchar *ent_name;
+
+	dir = g_dir_open (i2c_device_dir, 0, error);
+	if (dir == NULL) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "%s failed to open directory: %s.",
+			     __func__,
+			     i2c_device_dir);
+		return FALSE;
+	}
+
+	while ((ent_name = g_dir_read_name (dir)) != NULL) {
+		if (strlen (ent_name) <
+		    strlen (DEVICE_NAME_PREFIX) + HID_LENGTH)
+			continue;
+
+		if (add_i2c_device (plugin, i2c_device_dir,
+				    ent_name, error) == FALSE) {
+			return FALSE;
+		}
+	}
```

Comment:
This is really not the way to deal with i2c devices; you want to use FuUdevDevice and just ask fwupd to enumerate the correct sysfs class. e.g. see the WIP https://github.com/fwupd/fwupd/blob/wip/hughsie/elantp/plugins/elantp/fu-plugin-elantp.c and the other files in https://github.com/fwupd/fwupd/tree/wip/hughsie/elantp/plugins/elantp